pandas/core/indexes/base.py

```diff
@@ -2972,16 +2972,20 @@ def get_value(self, series, key):
         # use this, e.g. DatetimeIndex
         s = getattr(series, '_values', None)
         if isinstance(s, (ExtensionArray, Index)) and is_scalar(key):
-            # GH 20825
+            # GH 20882, 21257
             # Unify Index and ExtensionArray treatment
             # First try to convert the key to a location
-            # If that fails, see if key is an integer, and
+            # If that fails, raise a KeyError if an integer
+            # index, otherwise, see if key is an integer, and
             # try that
             try:
                 iloc = self.get_loc(key)
                 return s[iloc]
             except KeyError:
-                if is_integer(key):
+                if (len(self) > 0 and
+                        self.inferred_type in ['integer', 'boolean']):
```

Comment:
What makes this necessary?  I'm not clear on how it relates to the Mixin classes.

Reply:
For this you can comment on the relevant other PR (https://github.com/pandas-dev/pandas/pull/21260)

Reply:
@jbrockmendel the other PR (#21260) was needed to make the tests here work right.